src/enclave/enclave_oe.c

```diff
@@ -311,38 +305,87 @@ static void _read_eeid_config()
     sgxlkl_enclave_state.config = cfg;
 }
 
+void sgxlkl_ensure_inside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_within_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is not in enclave memory.\n");
+}
+
+void sgxlkl_ensure_outside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_outside_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is in enclave memory.\n");
+}
+
 static void _copy_shared_memory(const sgxlkl_shared_memory_t* host)
 {
     const sgxlkl_enclave_config_t* cfg = sgxlkl_enclave_state.config;
 
-    /* Deep copy where necessary */
+    /* The host shared memory struct has been copied by OE (but not its
+     * contents). */
+    sgxlkl_ensure_inside(host, sizeof(sgxlkl_shared_memory_t));
 
-    sgxlkl_shared_memory_t* enc = &sgxlkl_enclave_state.shared_memory;
-    memset(enc, 0, sizeof(sgxlkl_shared_memory_t));
+    /* Temporary, volatile copy to make sure checks aren't reordered */
+    volatile sgxlkl_shared_memory_t* enc = oe_calloc_or_die(
+        1,
+        sizeof(sgxlkl_shared_memory_t),
+        "Could not allocate memory for shared memory\n");
 
     if (cfg->io.network)
+    {
         enc->virtio_net_dev_mem = host->virtio_net_dev_mem;
+        sgxlkl_ensure_outside(
+            enc->virtio_net_dev_mem, sizeof(struct virtio_dev));
+    }
 
     if (cfg->io.console)
+    {
         enc->virtio_console_mem = host->virtio_console_mem;
+        sgxlkl_ensure_outside(
+            enc->virtio_console_mem, sizeof(struct virtio_dev));
+    }
+
+    size_t num_channels = host->num_evt_channel;
+    if (num_channels > 0)
+    {
+        _Static_assert(
+            sizeof(enc_dev_config_t) == 16UL,
+            "enc_dev_config_t size has changed");
+
+        _Static_assert(
+            sizeof(enc_dev_state_t) == 8UL, "enc_dev_state_t size has changed");
 
-    enc->evt_channel_num = host->evt_channel_num;
-    /* enc_dev_config is required to be outside the enclave */
-    enc->enc_dev_config = host->enc_dev_config;
+        enc->num_evt_channel = num_channels;
+        enc->enc_dev_config = oe_calloc_or_die(
+            num_channels,
+            sizeof(enc_dev_config_t),
+            "Could not allocate memory for event channel config\n");
+
+        for (size_t i = 0; i < num_channels; i++)
+        {
+            const enc_dev_config_t* host_conf_i = &host->enc_dev_config[i];
+            enc_dev_config_t* enc_conf_i = &enc->enc_dev_config[i];
+            enc_conf_i->dev_id = host_conf_i->dev_id;
+            enc_conf_i->enc_evt_chn = host_conf_i->enc_evt_chn;
+            sgxlkl_ensure_outside(
```

Comment:
Simply checking enc_evt_chn is outside of the enclave at this point is not enough, as the malicious host can change the host memory content after the check hear passes. Probably should analyze the data structure and keep a "shadow" structure inside the enclave and copy out enclave update to the host-read-only components from the "shadow " structure to the strcuture outside of the enclave.

Reply:
Yes, that's correct and expected; this check here is really redundant. The content of this data structure is shared memory that both host and enclave write to (all of the contents) and the important checks are in `enclave_event_channel.c`.